self checks

jlurgo opened the pull request:
- Performs self checks on agent startup
  - can use unshare
  - bash is installed and is the default shell ( only added to test having multiple checks)
- Self-check results are shown reactively in the Device details section

![image](https://github.com/user-attachments/assets/5decb97a-a09a-4481-b8be-869e7266f15a)
![image](https://github.com/user-attachments/assets/83a0c864-fd30-4972-a015-dec03d67122d)


Ref: https://github.com/transitiverobotics/issues/issues/400
closes: https://github.com/transitiverobotics/issues/issues/608

chfritz commented on robot-agent/mqtt.js:
```diff
@@ -215,6 +215,74 @@ const staticInfo = () => {
+  });
+  if (errors.length > 0) {
+    log.error('self checks failed:', errors);
+    data.update(`${AGENT_PREFIX}/status/selfChecks`, {errors});
+  } else {
+    log.info('self checks passed');
+    data.update(`${AGENT_PREFIX}/status/selfChecks`, {success: true});
```
As discussed, I think we should *not* send the error message, only the test result. The front-end should read the error message from a shared file based on the error ID or name. So what we send to the front-end should just be an object `{ [errorName]: errorResult, .. }`.

chfritz replied:
For example: 
`{ 'unshare': 'cannot run unshare -rm, please do ...', condition: (result) => result.trim() == 'root', command: 'unshare -rm whoami'}`

chfritz commented on cloud/app/web_components/self-check.jsx:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  return (
+    <Alert variant="success">
+      <strong>All self-checks passed successfully!</strong>
+    </Alert>
```
This is non-news to the user. No need to show.

chfritz commented on robot-agent/mqtt.js:
```diff
@@ -215,6 +215,74 @@ const staticInfo = () => {
 };
 
+// list of self checks to run
+const selfChecks = [
```
Since all these use an execSync (as intended), I think we should make this declarative, not procedural. Meaning that `selfChecks` should just list the commands, a function to evaluate the result. Also, it should be an object (with name as key), not an array. This way we'll be able to look up the tests from the front-end.

chfritz commented on cloud/app/web_components/robot-agent-device.jsx:
```diff
@@ -469,6 +470,9 @@ const Device = (props) => {
       </Fold>
     </div>
 
+    <div style={styles.row}>
```
only show if there are failed tests

chfritz reviewed (commented):
Definitely along the lines of what I was thinking. See comments.

chfritz commented:
> What I haven't done is having the checks/errors in a shared file between agent/frontend, how should I do that? a symlink? a library?

Can you first try the naive approach of just importing a shared file directly in the UI: `import {checks} from '../../robot-agent/selfCheck.js;` -- or similar. Not sure the esbuild bundler will allow it, but we can try.

jlurgo commented:
OK looks like it works!

jlurgo commented:
> OK looks like it works!

Well it doesn't...  I'm trying to make it work, already tested a symlink which didn't work... I'm trying to make esbuild to make a copy of the file at build time, still no luck.

jlurgo commented:
I refactored the self checks, looks much better now

chfritz commented:
> already tested a symlink which didn't work

Check the esbuild in utils-caps, there symlinks work

jlurgo commented:
OK I refactored the self checks in a single file, agent side, and sending the error messages to the frontend.
Also added geolocation info in MqttSync.
I'm showing Country, City in the device details section, just to test, I can remove it if you want but I thought it looks good.

![image](https://github.com/user-attachments/assets/25f11d06-3e87-4017-a698-0251907d3ba3)

jlurgo commented:
also added the check for overlayfs support, but not very confident about it

chfritz commented on robot-agent/selfChecks.js:
```diff
@@ -43,6 +43,17 @@ const selfChecks = {
+  overlayNotSupported: {
+    command: 'TRPACKAGE=@test_overlay/test ./unshare.sh whoami',
+    checkResult: (result) => {
+      return false;
```
I find it confusing for `false` to mean test success. Shouldn't it be the other way round, `true` meaning pass, and `false` meaning fail?

jlurgo replied:
Since the test name is negative, in this case: **overlayNotSupported: true** for FAIL makes sense. but I can change it, it's also confusing for me...

jlurgo replied:
done

chfritz replied:
Thanks. This looks great!

chfritz reviewed (approved):
Looks good. There are some smaller changes I want to make, especially regarding the use of `await` of the fetch of the geoip, where I want to further reduce any risks to the agent starting when offline or that geoip service not responding. But I'll do that in a follow-up before releasing to the field. Btw, why did you not use ipconfig.io for geoip? I get a more accurate geo location from that than the ip-api.com you ended up using.
